B-Sockets-ServerTesting/sockets-servertesting.md

```diff
@@ -0,0 +1,298 @@
+- Reads and writes to the other side
+- Message boundaries may not be preserved
+- nearly the same as `write()` and `read()`, but with socket-specific options
+
+A TCP client may use these functions as such:
+```c
+int fd = socket(...);
+connect(fd, ... /* server address */);
+
+// Communicate with the server by send()ing from and recv()ing to fd.
+
+close(fd);
+```
+
+And a TCP server:
+
+```c
+int serv_fd = socket(...);
+bind(serv_fd, ... /* server address */);
+listen(serv_fd, ... /* max pending connections */);
+
+// use an infinite loop, to continue accepting new incoming clients
+for (;;) {
+    int clnt_fd = accept(serv_fd, ...);
+
+    // Communicate with the client by send()ing from and recv()ing to
+    // clnt_fd, NOT serv_fd.
+
+    close(clnt_fd);
+}
+```
+
+### Listening socket vs connected socket
+![](img/listening-vs-connecting.png)
+
+To form a bidirectional channel between client and server, three sockets are used:
+- The server uses two sockets
+  - The listening socket, to accept incoming connections from a client
+  - The client socket, which is created when an incoming connection has been
+    `accept()`ed. 
+- The client uses one socket
+  - The `connect()`ing socket, which reaches out to the server. Once the
+  connection has been made, communication can be done between the server's client
+  socket and the client's connecting socket.
+
+### HTTP 1.0
+HTTP 1.0 is a protocol between a client, typically a web browser, and a server,
+typically a web server hosting files such as HTML. It is an outdated version of
+the HTTP protocol and simpler than newer versions.
+
+When visiting a website, a URL is specified in the following format:
+
+```
+http://example.com:80/index.html
+^^^^   ^^^^^^^^^^^ ^^ ^^^^^^^^^^
+|      |           |  |
+|      |           |  URI = /index.html
+|      |           port number = 80
+|      domain name = example.com
+protocol = HTTP
+```
+
+Based on the information provided by the user in the URL, a web client will
+establish a socket connection with the IP address of the domain name. After
+establishing the connection, the two computers exchange text in the form of HTTP
+requests:
+
+- The client sends an **HTTP request** for a resource on the server
+- The server sends an **HTTP response**
+
+**HTTP request**
+- First line: method, request URI, version
+    - Ex: "GET /index.html HTTP/1.0\r\n"
+- Followed by 0 or more headers
+    - Ex: "Host: www.google.com\r\n"
+- Followed by an empty line
+    - "\r\n"
+
+**HTTP response**
+- First line: response status
+    - Success: "HTTP/1.0 200 OK\r\n"
+    - Failure: "HTTP/1.0 404 Not Found\r\n"
+- Followed by 0 or more response headers
+- Followed by an empty line
+- Followed by the content of the response
+    - Ex: image file or HTML file
+
+We can see the contents of real HTTP requests using `netcat` by pretending to be
+either a web client or server. 
+
+Let's first act as a web server. We tell `netcat` to open a server connection
```

Comment:
maybe im tweaking, but i am worried that this example might confuse students. i think it's great to use nc -l to view how a client request comes in, but i am not sure if it is clear enough that nc cant actually be a real web server since it cant understand HTTP requests... like if a student tries this, their browser will just hang

Reply:
I added a sentence about it but lmk if you think there needs to be more